keps/sig-storage/1472/README.md

```diff
@@ -0,0 +1,1349 @@
+
+To ensure that `CSIStorageCapacity` objects get removed when the driver
+deployment gets removed before it has a chance to clean up, each
+`CSIStorageCapacity` object needs an [owner
+reference](https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#OwnerReference).
+
+For central provisioning, that has to be the Deployment or StatefulSet
+that defines the provisioner pods. That way, provisioning can
+continue seamlessly when there are multiple instances with leadership
+election.
+
+For deployment with a daemon set, making the individual pod the owner
+is better than the daemon set as owner because then other instances do
+not need to figure out when to remove `CSIStorageCapacity` objects
+created on another node. It is also better than the node as owner
+because a driver might no longer be needed on a node although the node
+continues to exist (for example, when deploying a driver to nodes is
+controlled via node labels).
+
+While external-provisioner runs, it needs to update and potentially
+delete `CSIStorageCapacity` objects:
+- when nodes change (for central provisioning)
+- when storage classes change (for persistent volumes)
+- when volumes were created or deleted (for central provisioning)
+- when volumes are resized or snapshots are created or deleted (for persistent volumes)
+- periodically, to detect changes in the underlying backing store (all cases)
+
+Because sidecars are currently separated, external-provisioner is
+unaware of resizing and snapshotting. The periodic polling will catch up
+with changes caused by those operations.
+
+### Using capacity information
+
+The Kubernetes scheduler already has a component, the [volume
+scheduling
+library](https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/volume/scheduling),
+which implements [topology-aware
+scheduling](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/volume-topology-scheduling.md).
+
+The
+[CheckVolumeBinding](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/volume-topology-scheduling.md#integrating-volume-binding-with-pod-scheduling)
+function gets extended to not only check for a compatible topology of
+a node (as it does now), but also to verify whether the node falls
+into a topology segment that has enough capacity left. This check is
+only necessary for PVCs that have not been bound yet.
+
+The lookup sequence will be:
+- find the `CSIDriver` object for the driver
+- check whether it has `CSIDriver.spec.storageCapacity` enabled
+- find all `CSIStorageCapacity` objects that have the right spec
+  (driver, accessible by node, storage class) and sufficient capacity.
+
+The specified volume size is compared against `Capacity` if
+available. A topology segment which has no reported capacity or a
+capacity that is too small is considered unusable at the moment and
+ignored.
+
+Each volume gets checked separately, independently of other volumes
```

Comment:
As a best effort, can we at least sum up the capacities for multiple volumes in the same Pod? This can at least reduce the chance of running into partial binding failures.

As a local volume example, if the node says it has "50GB" left, and a pod requests two local volumes of "50GB" each, then we should be fairly confident that the pod definitely cannot fit.

Reply:
That requires summing up the capacity of all the requested volumes and making a decision at the same time?  But the create volume requests may not all come at exactly the same time.

Reply:
> As a local volume example, if the node says it has "50GB" left, and a pod requests two local volumes of "50GB" each, then we should be fairly confident that the pod definitely cannot fit.

This makes assumptions about the semantic of "capacity", something that we wanted to avoid. It also assumes that "capacity" == "total capacity". It's erring on the wrong side if the driver reports "maximum volume size".

Let's first build something that doesn't use heuristics and then check how well it works before introducing additional logic, okay?